### PR TITLE
[Merged by Bors] - feat: Add Uncountable

### PR DESCRIPTION
**Pull request by JADekker**
Adds the Uncountable class to `Countable/Defs` and some basic API.

---
 More to follow, but I'd like to get feedback on the way of writing first! 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alexjbest** [comment]
Do you have a link to the other things you want to add?

File: Mathlib/Data/Countable/Defs.lean [hunk 1]
```diff
@@ -117,3 +119,58 @@ instance (priority := 500) Quotient.countable [Countable α] {r : α → α →
 
 instance (priority := 500) [Countable α] {s : Setoid α} : Countable (Quotient s) :=
   (inferInstance : Countable (@Quot α _))
+
+/-
```
Comment (alexjbest) [inline Mathlib/Data/Countable/Defs.lean:123]: Could this be a module comment so it shows up in docs

**Comment by JADekker** [comment]
Thanks, I’ll make the adjustments soon.

short background:

the reason why I’m implementing this is that I’m working on Lindelöf spaces in #9107. I’d like to introduce the cocountable filter (analogous to the way the cofinite filter is used for Compact spaces) and it came up in discussions on Zulip that, rather than writing \neg Countable at some points in the definitions, it might be nicer to have Uncountable. 

As such I’ll mirror the contents from Countable/Basic for Uncountable, but after that I’ll probably only add whatever comes up that I need to define cocountable filters (but I’ll try to prove and add sufficiently general versions of whatever I’m using).

**Comment by urkud** [comment]
I'm done working on this branch. I think that this PR needs the instance for functions I posted on Zulip and the docstring fix suggested by Alex; otherwise it's ready. The instance should probably go to `Data/Fintype/Card`, because you'll need `Infinite.natEmbedding`.

**Comment by JADekker** [comment]
Great, will try to do this in a few days; if anyone has time and wants to add them, feel free of course! I’ll send a heads-up when I continue working on.

**Comment by JADekker** [comment]
@urkud Do you mean the instance 
[Uncountable α] : NeBot (cocountable α)
that you suggested in Zulip, or do you mean another one? This one requires the definition of the cocountable filter from #9200 which is still pending, and requires filters, so I think Data/Fintype/Card may be too far upstream (as I'll need to get NeBot, which comes from Filter/Basic). Perhaps I should include it in the PR for Filter/Cocountable when this PR gets merged?

File: Mathlib/Data/Countable/Basic.lean [hunk 2]
```diff
@@ -43,6 +46,9 @@ protected theorem Function.Embedding.countable [Countable β] (f : α ↪ β) :
   f.injective.countable
 #align function.embedding.countable Function.Embedding.countable
 
+protected lemma Function.Embddding.uncountable [Uncountable α] (f : α ↪ β) : Uncountable β :=
```
Comment (YaelDillies) [inline Mathlib/Data/Countable/Basic.lean:49]:
```suggestion
protected lemma Function.Embedding.uncountable [Uncountable α] (f : α ↪ β) : Uncountable β :=
```

File: Mathlib/Data/Countable/Basic.lean [hunk 3]
```diff
@@ -53,24 +59,51 @@ section type
...
+instance Sum.uncountable_inl [Uncountable α] : Uncountable (α ⊕ β) :=
+  inl_injective.uncountable
+
+instance Sum.uncountable_inr [Uncountable β] : Uncountable (α ⊕ β) :=
+  inr_injective.uncountable
+
 instance [Countable α] : Countable (Option α) :=
   Countable.of_equiv _ (Equiv.optionEquivSumPUnit.{_, 0} α).symm
 
+instance [Uncountable α] : Uncountable (Option α) :=
```
Comment (YaelDillies) [inline Mathlib/Data/Countable/Basic.lean:76]:
```suggestion
instance Option.instUncountable [Uncountable α] : Uncountable (Option α) :=
```
and elsewhere

File: Mathlib/Data/Countable/Basic.lean [hunk 4]
```diff
@@ -53,24 +59,51 @@ section type
...
+  Injective.uncountable fun _ _ ↦ Option.some_inj.1
```
Comment (YaelDillies) [inline Mathlib/Data/Countable/Basic.lean:77]:
```suggestion
  some_injective.uncountable
```

File: Mathlib/Data/Countable/Defs.lean [hunk 5]
```diff
@@ -117,3 +119,59 @@ instance (priority := 500) Quotient.countable [Countable α] {r : α → α →
...
+  not_countable : ¬Countable α
+
+lemma not_uncountable_iff : ¬Uncountable α ↔ Countable α := by
+  rw [uncountable_iff_not_countable, not_not]
+
+lemma not_countable_iff : ¬Countable α ↔ Uncountable α := (uncountable_iff_not_countable α).symm
+
+@[simp]
+lemma not_uncountable [Countable α] : ¬Uncountable α :=
+  mt not_countable_iff.2 <| not_not_intro ‹_›
```
Comment (YaelDillies) [inline Mathlib/Data/Countable/Defs.lean:140]:
```suggestion
lemma not_uncountable [Countable α] : ¬Uncountable α := not_uncountable_iff.2 ‹_›
```

File: Mathlib/Data/Countable/Defs.lean [hunk 6]
```diff
@@ -117,3 +119,59 @@ instance (priority := 500) Quotient.countable [Countable α] {r : α → α →
...
+protected theorem Function.Injective.uncountable [Uncountable α] {f : α → β} (hf : Injective f) :
+    Uncountable β :=
+  ⟨fun _ ↦ not_countable hf.countable⟩
+
+protected theorem Function.Surjective.uncountable [Uncountable β] {f : α → β} (hf : Surjective f) :
+    Uncountable α := (injective_surjInv hf).uncountable
+
+lemma not_injective_uncountable_countable [Uncountable α] [Countable β] (f : α → β) :
+    ¬Injective f := fun hf ↦
+  not_countable hf.countable
```
Comment (YaelDillies) [inline Mathlib/Data/Countable/Defs.lean:154]:
```suggestion
    ¬Injective f := fun hf ↦ not_countable hf.countable
```
and elsewhere

File: Mathlib/Data/Countable/Basic.lean [hunk 7]
```diff
@@ -114,8 +114,8 @@ variable {α : Sort u} {β : Sort v} {π : α → Sort w}
 ### Operations on `Sort*`s
 -/
 
-instance [Countable α] [Countable β] : Countable (PSum α β) :=
-  Countable.of_equiv (Sum (PLift α) (PLift β)) (Equiv.plift.sumPSum Equiv.plift)
+instance [Countable α] [Countable β] : Countable (α ⊕' β) :=
+  Countable.of_equiv ((PLift α) ⊕ (PLift β)) (Equiv.plift.sumPSum Equiv.plift)
```
Comment (YaelDillies) [inline Mathlib/Data/Countable/Basic.lean:118]:
```suggestion
  Countable.of_equiv (PLift α ⊕ PLift β) (Equiv.plift.sumPSum Equiv.plift)
```

Comment (JADekker) [reply Mathlib/Data/Countable/Basic.lean:118]: We were doing the same thing I see! In Basic.lean, the instance Option.instUncountable gives an error for me, as it does not know some_injective.uncountable. Any clues?

File: Mathlib/Data/Countable/Basic.lean (hunk 4 above)
Comment (JADekker) [reply Mathlib/Data/Countable/Basic.lean:77]: That does not seem to work for me, I reverted this change!

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**